doc: Sync classref with current source

akien-mga opened the pull request:
Needs review/amends, some of those changes might not be fully intended.

akien-mga commented on doc/classes/ArrayMesh.xml:
```diff
@@ -272,14 +298,44 @@
+		<constant name="ARRAY_FORMAT_INDEX" value="4096" enum="ArrayFormat">
 			Index array will be used.
 		</constant>
+		<constant name="ARRAY_FORMAT_BLEND_SHAPE_MASK" value="-8185" enum="ArrayFormat">
```
@reduz Overflowing `int32_t`?

akien-mga replied:
Same stuff on `Mesh`. BTW shouldn't we bind those only in `Mesh` and not `ArrayMesh`? Aren't they exactly the same?

akien-mga replied:
> Overflowing `int32_t`?

reduz will fix that in a follow-up PR.

> BTW shouldn't we bind those only in Mesh and not ArrayMesh? Aren't they exactly the same?

I fixed that.

akien-mga commented on doc/classes/ArrayMesh.xml:
```diff
@@ -272,14 +298,44 @@
 		</constant>
+		<constant name="ARRAY_FORMAT_BLEND_SHAPE_MASK" value="-8185" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FORMAT_CUSTOM_BASE" value="13" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FORMAT_CUSTOM0_SHIFT" value="13" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FORMAT_CUSTOM1_SHIFT" value="16" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FORMAT_CUSTOM2_SHIFT" value="19" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FORMAT_CUSTOM3_SHIFT" value="22" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FORMAT_CUSTOM_MASK" value="7" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_COMPRESS_FLAGS_BASE" value="25" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FLAG_USE_2D_VERTICES" value="33554432" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FLAG_USE_DYNAMIC_UPDATE" value="67108864" enum="ArrayFormat">
+		</constant>
+		<constant name="ARRAY_FLAG_USE_8_BONE_WEIGHTS" value="134217728" enum="ArrayFormat">
+		</constant>
```
All these values are pretty weird. Is that intended?

akien-mga replied:
Seems intended. I'm not sure the `_BASE` variables are meant to be exposed though.

akien-mga commented on doc/classes/Control.xml:
```diff
@@ -133,6 +133,14 @@
 				Return [code]Array[/code] of [code]Vector2i[/code] text ranges, in the left-to-right order. Ranges should cover full source [code]text[/code] without overlaps. BiDi algorithm will be used on each range separately.
 			</description>
 		</method>
+		<method name="_unhandled_key_input" qualifiers="virtual">
```
That's already bound in `Node`, why is it now re-bound in `Control` (and with wrong type hint)?

Comes from efe5c250d5c @EricEzaM.

EricEzaM replied:
First guess would be that I didn't bind it as virtual? In mobile can't check atm

akien-mga replied:
Why does it need to be bound though? It's already bound in Node, and Control inherits Node.

EricEzaM replied:
Yeah that too lol

EricEzaM replied:
I'll fix it up when I get to my computer

akien-mga replied:
Fixed.

akien-mga commented on doc/classes/RenderingServer.xml:
```diff
@@ -3084,16 +3068,16 @@
 			Array is an UV coordinates array for the second UV coordinates.
 		</constant>
-		<constant name="ARRAY_BONES" value="6" enum="ArrayType">
+		<constant name="ARRAY_BONES" value="10" enum="ArrayType">
```
Missing intermediate bindings?

akien-mga replied:
Fixed. You systematically forget to bind enum values in `rasterizer_server.cpp` :)

akien-mga commented on doc/classes/RenderingServer.xml:
```diff
@@ -3114,40 +3098,19 @@
 			Flag used to mark an UV coordinates array for the second UV coordinates.
 		</constant>
-		<constant name="ARRAY_FORMAT_BONES" value="64" enum="ArrayFormat">
+		<constant name="ARRAY_FORMAT_BONES" value="1024" enum="ArrayFormat">
```
Missing bindings.

akien-mga replied:
Fixed.

akien-mga commented on doc/classes/RichTextLabel.xml:
```diff
@@ -329,9 +329,6 @@
 			The raw text of the label.
 			When set, clears the tag stack and adds a raw text tag to the top of it. Does not parse BBCodes. Does not modify [member bbcode_text].
 		</member>
-		<member name="text_direction" type="int" setter="set_text_direction" getter="get_text_direction" enum="Control.TextDirection" default="0">
-			Base text writing direction.
-		</member>
```
I guess this was a rebase mistake and is actually added by #42595 @bruvzg?

bruvzg replied:
Yes, it should be in #42595.

akien-mga commented on doc/classes/SurfaceTool.xml:
```diff
@@ -5,14 +5,15 @@
 	</brief_description>
 	<description>
 		The [SurfaceTool] is used to construct a [Mesh] by specifying vertex attributes individually. It can be used to construct a [Mesh] from a script. All properties except indices need to be added before calling [method add_vertex]. For example, to add vertex colors and UVs:
+		[b]FIXME: Code no longer valid in 4.0, add_uv and add_color were removed. What's the replacement?[/b]
```
Should we just replace the `add_` methods by `set_` methods now? @reduz

akien-mga replied:
Did that.

akien-mga commented on doc/classes/SurfaceTool.xml:
```diff
@@ -167,11 +104,11 @@
 			<description>
 				Returns a constructed [ArrayMesh] from current information passed in. If an existing [ArrayMesh] is passed in as an argument, will add an extra surface to the existing [ArrayMesh].
-				Default flag is [constant Mesh.ARRAY_COMPRESS_DEFAULT]. See [code]ARRAY_COMPRESS_*[/code] constants in [enum Mesh.ArrayFormat] for other flags.
+				See [code]ARRAY_COMPRESS_*[/code] constants in [enum Mesh.ArrayFormat] for possible flags.
```
What should this be updated to? There's only one `ARRAY_COMPRESS_*` flag left.

akien-mga replied:
Added a FIXME.

akien-mga committed:
doc: Sync classref with current source

And fixups:
- Add missing bindings in RenderingServer
- Remove duplicate ArrayMesh enum bindings (they're in Mesh already)
- Remove redundant _unhandled_key_input binding in Control (it's in Node
  already)

akien-mga commented:
Removed the redundant bindings in `ArrayMesh` (already in `Mesh`) as advised by @reduz.

aaronfranke commented:
> @akien-mga: Removed the redundant bindings in `ArrayMesh` (already in `Mesh`) as advised by @reduz.

See also #36382, specifically commit "Rebind Mesh/ArrayMesh enums" which removes the redundant bindings in `ArrayMesh`, but unlike here, it combines the docs by keeping the added detail currently present in the `ArrayMesh` docs.

akien-mga commented:
> See also #36382, specifically commit "Rebind Mesh/ArrayMesh enums" which removes the redundant bindings in ArrayMesh, but unlike here, it combines the docs by keeping the added detail currently present in the ArrayMesh docs.

Seems good, though this commit also removes more duplicated enum bindings in ArrayMesh, and fixes the code examples accordingly. So if that's OK with you I'd propose to merge this PR and you can include the additional BlendShapeMode and documentation changes while rebasing #36382?
